### PR TITLE
Allow access to template containers

### PR DESCRIPTION
**Pull request by kardeiz**
and make it possible to interact with them before adding them to managed state.

Also reexports `tera` and `handlebars`.

See #64.

### CODE REVIEW

File: contrib/src/templates/mod.rs [hunk 1]
```diff
@@ -3,25 +3,37 @@ extern crate serde_json;
...
-use rocket::State;
+use rocket::{Config, State};
 use rocket::request::Request;
 use rocket::fairing::{Fairing, AdHoc};
 use rocket::response::{self, Content, Responder};
 use rocket::http::{ContentType, Status};
 
+pub use self::context::Context;
+
+
```
Comment (SergioBenitez) [inline contrib/src/templates/mod.rs]: Why is there so much empty space here? Just one line is fine.

**Comment by SergioBenitez** [comment]
If we're going to accept this, it's going to need to be thoroughly documented. How do you intend this to be used?

**Comment by milyin** [comment]
It's definitely necessary to give user access to handlebars object to add custom helpers. Without custom helpers handlebars engine is too limited.
May be it's possible to allow to register some callback and call it instead of "let hb = Handlebars::new()" here https://github.com/SergioBenitez/Rocket/blob/master/contrib/src/templates/handlebars_templates.rs#L12 ?

**Comment by SergioBenitez** [comment]
@milyin I was hoping to expose an engine-agnostic mechanism for adding template helpers so that you can add one helper that works for both Handlebars and Tera (and any engines in the future). As far as I can see, being able to add helpers are the primary and perhaps only benefit of this PR. Is that the case? If so, I'd like to implement the nicer solution even though it means delaying being able to do this further.

**Comment by maddin77** [comment]
What about [filters](https://tera.netlify.com/docs/templates/#filters) and [tests](https://tera.netlify.com/docs/templates/#tests) in tera?

**Comment by SergioBenitez** [comment]
Good point, @maddin77.

@kardeiz I'm still in need of documentation for this change. I don't know how you expect these changes will be used; nothing looks obvious to me.

**Comment by kardeiz** [comment]
I'm not sure if this is the best way to do this; having some kind of generator arg (e.g., `Fn() -> Handlebars`) to `Template::fairing` might be better, or some kind of callback for the user (e.g., `Fn(&mut Context)` or `Fn(&mut Handlebars)`. I just put this together as the simplest way to get access to the extra features of `Tera`.

In any case, I was using this PR like:

```rust
rocket::ignite()
    .mount("/", routes![])
    .attach(AdHoc::on_attach(|rocket| {
        match Context::initialize(Template::root(rocket.config())) {
            Some(mut ctxt) => {
                // you can add filters, tests, etc. here
                configure(&mut ctxt.engines.tera);
                Ok(rocket.manage(ctxt))
            },
            None => Err(rocket)
        }
    }))
```

**Comment by SergioBenitez** [comment]
I like the idea of a callback! What if we add another method through which the fairing can be created, say `custom`, that take an `F: FnMut(&mut Engines)` through which the user can do whatever they want with the enabled templating engines before Rocket launches?

**Comment by SergioBenitez** [comment]
Closing this out in hopes of a better solution.

**Comment by vishusandy** [comment]
Are there any plans to allow custom helpers?  I would really like to be able to use them without having to create a local copy of rocket and modify the handlebars code (which I did to get it to work but it creates problems of its own).

**Pull request closed without merging**